### DISCUSSION TITLE
Completion of the View Transition API

### DISCUSSION THREAD

**Discussion by martrapp** [withastro/roadmap#770] [Proposal]
### Body

# Summary

There are still a few side issues around the View Transition API that we've left out so far. Since we are now plan to release the new events and 4.0 and Christmas are also just around the corner, now seems to be a good time to finalize everything:

1. Same-origin policy vs. opening up navigation with CORS
2. Unified support for intra page navigation
3. *Edited 23/11/17:* Just a placeholder for now: Error handling, interrupted / aborted navigation, `navigate()` returning promises,...
4. *Edited 23/11/18:* New topic added: Support API for custom `loader()` and `swap()`
5. **&lt;Add your points here>**

## Same-origin policy vs opening navigation up with CORS
Status quo:
* Click handler enforces same-origin policy
* navigate() does not. With the right CORS settings you can fetch the new DOM from where you want.

PRO of cross site navigation:
- Seamless navigation to e.g. own sub-domains (https://storefront.my.store -> https://catalog.my.store)

CON: 
- When migrating Astro view transitions to an implementation based on the Navigation API (which would have maintenance benefits), we can only support same-origin. The Navigation API enforces a same-origin policy, mainly because it provides readable history entries and doesn't want to expose data from other origins.

My opinion on this topic: *I'd love to hear feedback on whether smooth cross-origin navigation is a thing or if no one cares. It would be an impediment to a future move to the navigation API. Maybe we can find a trick to use redirects if people think this feature is important. Otherwise, we should just fall back to the "same-origin" policy now.*

## Unified support also for navigation within a page
The API is called Astro View Transitions, but the visual transition from one page to another is only one facet of what we are talking about here. The ingredients are:

A. Unified navigation action, triggered by links, forms, browser interaction for history navigation or programmatically by `navigate()` or `history.*()`

B. Updating the current DOM to resemble a new view (without hard reloading the page)

C. Animated visual transition from the old to the new view

D. Adding a new location to the browsing history (or replacing the current one)

E. Setting the scrollX and scrollY position according to the target URL

### Status Quo:

A) ~~works just fine. Edited 23/11/18: Oops, found one: `<area>` elements of `<map>` elements should also trigger `navigate()`~~ Done. See below.

B) Currently, the DOM is not updated if the <br>"`from` and `to` URLs have the same origin, pathname and search parameter values unless `options.formData` is set".<br>
*We plan to change this to* <br>
"`from` and `to` URLs have the same origin, pathname and search parameter values, unless `to.hash === ''"`

*If we exclude cross-origin navigation in the first question, we of course do not need to compare the origins.*

If the URLs "from" and "to" differ only by the fragment identifier (#hash), the page is scrolled and the history is updated, but nothing else happens. 
No events are generated and there is therefore no way for the user code to influence this behavior. For example, you can't use navigation to a # fragment on the current page as a trigger for a count or measurement or styling change or whatever. There is currently no way to achieve the same behavior (including events) as navigating to other pages or to the current page via a link without a # fragment. 

C) It is possible to skip visual transitions by setting `transition:animate=none` (or for fallback `<ViewTransitions fallback=swap/>`) and still have seamless navigation without hard page refreshes, just without animations. But you can't opt-in for animations when navigating to hash targets on the same page.

Added on 12/1: Here is a recipe to get intra-page view transitions with the current implementation:
- add an search parameter, e.g. `?vt` to the intra-page URL, thus forcing view transitions and events.
- in an `astro:before-preparation` listener: redefine the `loader()` to do nothing
- in an `astro:before-swap` listener: redefine swap() to do nothing (besides removing old instances of `.astro-route-announcer`)
- then choose one of the following two alternatives also for the `astro-before-swap` listener:
   - remove the search parameter added before from `e.to.search`. 
   - give the search parameter added above a random value in `e.to.search`.

The first alternative has the benefit, that the artificially introduced parameter does not show up in the browsers address bar. The second alternative has the benefit, that you also get intra-page view transitions when traversing back and forth through the browser history. 

D) This is currently the same behavior as having no view transitions: All navigations add a new entry to the browsing history (or replace the current entry, if desired) unless the URL of the new entry is the same as the current URL. It does not matter whether the target is on the current page or on another page. In the current implementation, this happens immediately before the `astro:after-swap` event is triggered.

E) This is currently the same behavior as when you have no view transition: If you land on a different page, you start at the beginning. If the "to" URL has a fragment identifier (#hash), you scroll down to it. Smooth scrolling is controlled by CSS. In the current implementation, this happens immediately before the "astro:after-swap" event is triggered. Currently there is no way to disable scrolling. 

So my take on these would be:

a) ~~some way to opt in for the full navigation processing (B) for navigation to a #hash fragment on the current page~~

b) ~~some way to opt in on view transitions (C) for navigation to a #hash fragment on the current page~~

c) no change for history state management (D) (✓)

d) ~~some way to opt out of scrolling (E).~~ Not required, see below 

Rational for a):
When we startet view transition we didn't need to deal with intra page links and that simplified the implementation. Now we have everything in place. But we must say: You can call `navigate()` but only some navigation will trigger events and allow you to hook in.

Rational for b):
I think the default should be that no view transitions are visible on these navigations. It would be irritating if the page fades, pops back up and then scrolls down.
But people may have great ideas on what to do if you can turn on view transitions for in-page navigations. Especially with the default group animation that moves and scales rectangles present on both views. Or parts of the page falling down, while smooth scrolling tries to catch up.

Rational for c):
There is currently no way to disable history state updates, and frankly I don't see the need for it. Also, if we want to switch the implementation to the Navigation API, we cannot disable history processing as it is an integral part of the Navigation API. The Navigation API sets the history state right at the beginning of the transition. Alternatively, you can choose to change the history at the end of the transition, which would match our current approach.

Rational for d):
When opted out, you can always scroll manually from the `astro:after-swap` event. I think you can develop interesting applications if you can keep or change the current position without being disrupted by the navigation. Imagine a series of pages with a clear structure that you can cycle through in the middle of the screen without having to jump back up and down between pages as you navigate and compare the different contents. I also would like to point out that the navigation API provides explicit control over scrolling ;-)

***What do you people think?***

**Comment by matthewp** [comment]
# Same-origin policy 

In general we've tried to keep `<ViewTransitions />` compatible with `<meta name="view-transition" content="same-origin" />`, so since that one is same origin ours should be too.

**Reply by martrapp** [reply]
Ah, very important point! I forgot about the future :)

**Comment by martrapp** [comment]
# New Topic: Support API for `loader()` and `swap()`
When you implement your own `loader()` and `swap()` you might appreciate it to get some support from an API. What part of the internal code of `defaultLoader()` and `defaultSwap()` should we export via `astro:transitions/client` so that custom code doesn't have to completely reinvent the wheel?

## Considerations for `loader()`

There are use cases where loader should be completely replaced, but I have seen nearly no examples of how to reuse parts of the default implementation. Originally, I thought it would be great to replace fetch so that it generated events with percentages of loading progress. But this is far less attractive than I thought:
1. streaming doesn't provide `content-length` headers (for good reasons)
2. if you set the `content-length` header explicitly for static files, platforms like Cloudflare explicitly throw it away as soon as they turn on compression. You can circumvent this with some custom header like `x-uncompressed-content-length`, but then &hellip;
3. &hellip; static files are loaded quickly. The only feedback is usually at the end: 100% done. 

Therefore, I see little reason to offer very fine-grained loading.

### Suggestion for `loader()` support:

``` javascript
loadDOMfromURL(event)
loadStyleLinks(newDocument)
loadClientOnlyStyles(newDocument, toURL)
```
The execution of these functions in the specified order corresponds to the execution of the standard implementation of `loader()`.

## Thoughts on `swap()`:

If someone wants to do swap completely differently, they probably don't need any support. But if you just want to redo the body swap, you'll be happy if you can reuse the rest  

### Suggestion for `swap()` support:

``` javascript
swapRootAttributes(newDocument)
swapScripts(newDocument)
swapHead(newDocument)
swapBody(newDocument)
swapPersisted(newDocument)
```
The execution of these functions in the specified order corresponds to the execution of the standard implementation of `swap()`.

**Comments on this?**

**Comment by martrapp** [comment]
`<area>` elements of `<map>` elements (imagemaps) should also trigger `navigate()`
(https://github.com/withastro/astro/issues/9001 already suggested that SVGAElements should be handled same way as HTMLAnchorElements)

I am in favor of treating both elements, HTMLAreaElement and SVGAElement in the click handler in the same way as HTMLAnchorElement

**Reply by martrapp** [reply]
This one was fixed with #9140

**Comment by martrapp** [comment]
> d) a way to disable scrolling (E).

I just realized that it is easy to disable scrolling by saving the current scroll position in the `astro:after-preparation` listener and restoring it in the `astro:after-swap` handler.  So we don't need an explicite option to opt-out from scrolling.

**Reply by kevintakeda** [reply]
Hey @martrapp, I noticed that scrolling occurs inside the `startViewTransition`, so the scroll animation and element animation happen simultaneously, resulting in a messy animation when more scrolling is required. The only fix seems to be resetting the scroll to 0 in the `astro:after-preparation` and then restoring the scroll manually when navigating backward, both in the `astro:after-preparation` and `astro:after-swap`.

Here is the workaround code: https://stackblitz.com/edit/github-teousd?file=src%2Flayout.astro
To see the issue, take off the event listeners, and try to navigate after scrolling down. I'm not sure if the issue happens only on my browser (Chrome 120). You can better see the issue by slowing down the animations on Chrome DevTools.

The solution I provided isn't 100% ideal since the user will see a flash of a scroll position on the old page before the transition occurs. I tried `scroll-behavior: auto;`, but it doesn't work. I'm investigating new solutions.

**Reply by kevintakeda** [reply]
The other solution I found is to give a unique name to each transition for every page. This makes each transition independent of the another, preventing scrolling animation between navigations. However, there are a few issues: 1) If the page uses 'searchParams,' it will still animate the scroll. 2) The 'new' animation will work, but the 'old' animation won't. 3) In case you give a dynamic name for the transition and use `searchParams`, the scroll animation still happens, and elements don't transition at all.

**Reply by martrapp** [reply]
Hi @mishimalisa thank you for your message!

# TL;DR:

Try this: remove all scripts and instead add this:

```
<style is:global>
  ::view-transition-group(*) {
    animation: none;
  }
  ::view-transition-old(root),
  ::view-transition-new(root) {
    animation: none;
  }
</style>
```

# Longer version

What you understandably think is scrolling is not scrolling, but the morph animation of the view transition. You have mitigated the perceived negative effects through clever JavaScript programming, but the more direct solution is to use suitable CSS definitions. This is a point that requires a lot of explanation, so here is a slightly longer answer.

First of all, each use of `transition:name` (direct or indirect via transition:animate) defines three different animations: An `old` animation that removes the old image and an `new` animation that blends in the new image. Both are typically controlled by Astro. The third animation is a group or morph animation defined by the browser.

## How the morph animation works:
If there are elements with the same view transition name on the outgoing and incoming page, the browser generates an animation in which it gradually changes the size and position of a rectangle with the old image so that it finally covers the rectangle of the new image. the new image lies underneath with the same change. During the animation, the upper old image slowly becomes invisible and the underlying new image appears.

If this animation is applied to large parts of the page, it actually looks like scrolling, but you can also see that the scrollbar does not move at all but has already jumped to the target position at the beginning of the animation.

## What animations are there?

By using `transition:name=none` you were actually able to eliminate some unwanted animations. But not the morph animations. It is probably not directly recognizable how many animations were actually defined in your example 😏.  Let me visualize it. The screenshots show the output of the [`<VtBotDebug>`](https://events-3bg.pages.dev/library/VtBotDebug/) component of the 👜Bag of Tricks✨ for Astro's View Transitions, [link to npm](https://www.npmjs.com/package/astro-vtbot)

These are the pseudo elements that the browser insert during view transitions:
![image](https://github.com/withastro/roadmap/assets/94928215/4cdddb22-0cc5-437c-bb6d-200d19d61010)

As you haven't explicitly set `transition:name` attributes, Astro generated some random names.
Here is the mapping to your DOM:
![image](https://github.com/withastro/roadmap/assets/94928215/fc4105b9-5625-44bc-a37e-0c804f8f36aa)

And these are the animations that actually ran on the view transition:
![image](https://github.com/withastro/roadmap/assets/94928215/5dd5924b-17ed-4c11-b65c-f548bc9aa1c1)

Transition 2, which holds your color elements, has all three kinds of animations:
The a) browser generated group animation and b) old and c) new animations that come from Astro's slide animation. 

For transition 3 (body) and transition 4 (nav) you removed the `new` and `old` animations by setting `transition:animate=none`. But the browser generated group animations are still there.

And there is also the browser induced root pseudo tree that is connected with the document's `<html>` element and also provides three animations (all browser defined).

## How to get rid of "scrolling"?
The animation that disturbs you is the group animation on the body. Tt morphs the `<body>` of the scrolled down page onto the `<body>` of the next page which after loading shows on vertical scroll position 0. This morphing is what looks like scrolling.
  
This particular morph animation could easily be removed: Just remove the transition:animate="none" property from the body. Then everything that has `astro-c5xlmjkn-1` goes away. But then there would still be the `scrolling` group animation on the `<html>` element, which just has the same visual effect. 

The CSS from the beginning of this answer (be sure to set `is:global`, scoped CCS will not work for pseudo elements) removes **all** group animations and the default old and new animations of the root = `<html>` element.

What remains is this:
![image](https://github.com/withastro/roadmap/assets/94928215/40fbb5df-1c1b-4071-9d9b-322675e20848)

When you scroll down, navigate forward, scroll down, and navigate back again you should see the slide effect without any "smooth scrolling".

I'm curious to see if it works for you too.

**Reply by kevintakeda** [reply]
Big thanks for the detailed answer! It truly helped me understand better how the View Transition API works. Your solution definitely fixes the issue I was trying to solve, which I naively thought was a "scroll animation" issue.

**Reply by martrapp** [reply]
Good to hear! You are welcome!

**Comment by martrapp** [comment]
# Progress Report

Most of the questions asked above are now answered: We do not need extensions for intra-page view transitions or to disable scrolling. The router will only support same-origin navigation, and we apply a small correction to what kinds of navigation will trigger view transitions.  

We will support full view transitions  
- when `from` and `to` have same origin (no CORS). Otherwise the `to` location is loaded via a normal browser navigation.
- when `from` and `to` URLs differ in `pathname` or `search`. If `from`and `to` only differ in the `hash` value (pure hash change) that only triggers view transitions if `to.hash === ''`. This mimics browser behavior if view transitions are not enabled and also supports view transitions for form POSTs to the current location. For all other pure hash changes, the scroll position, the future `from` location, and the history entry are updated, but no events are fired and there are no view transitions. See note on 'C)' in the first entry on the top of this page on how to force intra-page view transitions using `search` parameters. 

Which parts of the standard implementation of `loader()` and `swap()` should be exported to users is open. So far we have no _real_ requests for this. I would like to postpone the decision until we have _real_ evidence that there is a _real_ need for this. Really :)

**Comment by ktym4a** [comment]
I would like to discuss about the trigger rule for view transitions.

First of all I gave two names.
1. `Non-hash based trigger` (Current behavior)
2. `Request based trigger`

## `Non-hash based trigger`

| Action | From | To   | startViewTransition | Request |
|--------|------|------|---------------------|---------|
| Back  | ''   | ''   | Call                 | Has     |
| Back  | ''   | Hash | No                   | Has     |
| Back  | Hash | ''   | Call                | No     |
| Back  | Hash | Hash | No                  | No      |
| Forward  | ''   | ''   | Call                | Has     |
| Forward  | ''   | Hash | No                  | No      |
| Forward  | Hash | ''   | Call                | Has     |
| Forward  | Hash | Hash | No                  | No      |

To me this feels inconsistent.
Because when I go from '' to hash the View Transition do not start, but when I go back it start.

so I propose `Request based trigger`.

## `Request based trigger`
| Action | From | To   | startViewTransition | Request |
|--------|------|------|---------------------|---------|
| Back  | ''   | ''   | Call                 | Has     |
| Back  | ''   | Hash | Call                   | Has     |
| Back  | Hash | ''   | no                | No     |
| Back  | Hash | Hash | no                  | No      |
| Forward  | ''   | ''   | Call                | Has     |
| Forward  | ''   | Hash | no                  | No      |
| Forward  | Hash | ''   | Call                | Has     |
| Forward  | Hash | Hash | no                  | No      |

This one, If go back to the page where the View Transition started, the View Transition will start the same way and be more consistent.

Either way, it would be good to have a table like this in the docs.

**Reply by martrapp** [reply]
Hi @ktym4a, thanks for pointing this out!

The rule Astro tries to follow is this: If there would be a full page load when running without view transitions, there will be a view transition when you add `<ViewTransition />`.

I guess this is what the forward case does and where both schemes do not differ.

When I interpret the second scheme correctly, the difference would be in the back navigation when one URL has a hash fragment and the other doesn't, but both point to the same page. 

And in deed you have a point here: Without view transitions, the browser reloads on back navigation from the plain URL to the #hash URL.   

I will take a look into two things:
- how feasible would it be to change the implementation
- how can we make this a non braking change, e.g. by adding another parameter to `<ViewTransitions/>`

**Reply by ktym4a** [reply]
@martrapp Thanks and Your interpretation is correct.

If it is just changing the rule from the first to the second, I think I already know how to do it.
and I am willing to issue a PR if needed.

**Reply by martrapp** [reply]
A PR would be great💜
Please go ahead!

**Reply by martrapp** [reply]
The router is already well prepared to handle view transitions to any target URL including those with #links. Introducing the new behavior seems to be a one line change.
Please just make sure, that we have a switch to control it. I guess we should not change the default prior to Astro V5 as for some sites this might be a breaking change. Pulling @matthewp into the discussion for feedback at that point :) See also https://github.com/withastro/astro/pull/9991
